src/components/action-popover/action-popover-item/action-popover-item.component.tsx

```diff
@@ -124,42 +133,96 @@ export const ActionPopoverItem = ({
     "ActionPopoverItem only accepts submenu of type `ActionPopoverMenu`"
   );
 
-  const { setOpenPopover, isOpenPopover, focusButton } = context;
+  const {
+    setOpenPopover,
+    isOpenPopover,
+    focusButton,
+    submenuPosition,
+  } = context;
   const isHref = !!href;
   const [containerPosition, setContainerPosition] = useState<
     ContainerPosition | undefined
   >(undefined);
   const [guid] = useState(createGuid());
   const [isOpen, setOpen] = useState(false);
   const [focusIndex, setFocusIndex] = useState<number>(0);
-  const [isLeftAligned, setIsLeftAligned] = useState(true);
 
   const submenuRef = useRef<HTMLDivElement>(null);
   const ref = useRef<HTMLButtonElement>(null);
   const mouseEnterTimer = useRef<NodeJS.Timeout | null>(null);
   const mouseLeaveTimer = useRef<NodeJS.Timeout | null>(null);
+  const hasSubmenuAligned = useRef(false);
 
   useEffect(() => {
     if (!isOpenPopover) {
       setOpen(false);
     }
   }, [isOpenPopover]);
 
+  useEffect(() => {
+    if (icon) {
+      setChildHasIcon?.(true);
+    }
+    if (submenu) {
+      setChildHasSubmenu?.(true);
+    }
+  }, [icon, setChildHasSubmenu, setChildHasIcon, submenu]);
+
   const alignSubmenu = useCallback(() => {
-    if (checkRef(ref) && checkRef(submenuRef) && submenu) {
-      const align = leftAlignSubmenu(ref, submenuRef);
-      setIsLeftAligned(align);
-      setContainerPosition(getContainerPosition(ref, submenuRef, placement));
+    const checkCalculatedSubmenuPosition = calculateSubmenuPosition(
+      ref,
+      submenuRef,
+      submenuPosition,
+      currentSubmenuPosition
+    );
+
+    setCurrentSubmenuPosition?.(checkCalculatedSubmenuPosition);
+
+    return checkRef(ref) && checkRef(submenuRef) && submenu;
+  }, [
+    submenu,
+    setCurrentSubmenuPosition,
+    submenuPosition,
+    currentSubmenuPosition,
+  ]);
+
+  useEffect(() => {
+    function getContainerPosition(
+      itemRef: React.RefObject<HTMLElement>,
+      containerSubmenuRef: React.RefObject<HTMLElement>,
+      containerPlacement: "bottom" | "top"
+    ): ContainerPosition | undefined {
+      /* istanbul ignore if */
+      if (!itemRef.current || !containerSubmenuRef.current) return undefined;
+
+      const { offsetWidth: submenuWidth } = containerSubmenuRef.current;
+
+      const leftAlignedSubmenu = currentSubmenuPosition === "left";
+      const leftValue = leftAlignedSubmenu ? -submenuWidth : "auto";
+      const rightValue = leftAlignedSubmenu ? "auto" : -submenuWidth;
+      const yPositionName = containerPlacement === "top" ? "bottom" : "top";
+
+      return {
+        left: leftValue,
+        [yPositionName]: "calc(-1 * var(--spacing100))",
+        right: rightValue,
+      };
     }
-  }, [submenu, placement]);
+    setContainerPosition(getContainerPosition(ref, submenuRef, placement));
+  }, [submenu, placement, currentSubmenuPosition]);
 
   useEffect(() => {
-    alignSubmenu();
+    if (submenu && !hasSubmenuAligned.current) {
```

Comment:
**question** (non-blocking): I don't think it's at all likely to happen in practice, but just in case - I'm wondering what happens if the `submenu` prop gets dupdated/toggled during the component's lifecycle. The way you have it now, it looks like it will be fine the first time it comes on, even if that's well ater the first render. But if it gets toggled from off to on to off and then on again(!), then `alignSubmenu` will only be run the first time it turns on, which will set the ref flag to `true` and stop it being run the second time. Might this matter? (I honestly don't know - but it's such an edge case I doubt it's worth worrying much over...)

Reply:
Agree with Robin, worth seeing if this works with a deps array

Reply:
So I've tested this using a `setTimeout` to set the `submenu` prop to either a submenu or null, and it appears to work fine. Even when the default is `null`, and a submenu is passed lets say 5 seconds after render, the submenu is adjusted correctly.  Even when there is not enough space on the right for example, and then submenus are passed 5 seconds later, it inverts as expected. I'm happy to add this into a codeSandbox after the next batch of changes to show this, so you can both play around with it, but to me, it looks ok 👍